plugins/outputs/postgresql/column_details.go

```diff
@@ -0,0 +1,192 @@
+package postgresql
+
+import (
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/outputs/postgresql/utils"
+)
+
+// a function type that generates column details for the main, and tags table in the db
+type targetColumnInitializer func() (*utils.TargetColumns, *utils.TargetColumns)
+
+// constants used for populating the 'targetColumnInit' map (for better readability)
+const (
+	cTagsAsFK     = true
+	cTagsAsJSON   = true
+	cFieldsAsJSON = true
+)
+
+// Since some of the target columns for the tables in the database don't
+// depend on the metrics received, but on the plugin config, we can have
+// constant initializer functions. It is always known that the 'time'
+// column goes first in the main table, then if the tags are kept in a
+// separate table you need to add the 'tag_id' column...
+// This map contains an initializer for all the combinations
+// of (tagsAsFK, tagsAsJSON, fieldsAsJSON).
+var targetColumnInit = map[bool]map[bool]map[bool]targetColumnInitializer{
+	cTagsAsFK: {
+		cTagsAsJSON: {
+			cFieldsAsJSON:  tagsAsFKAndJSONAndFieldsAsJSONInit,
+			!cFieldsAsJSON: tagsAsFKAndJSONInit,
+		},
+		!cTagsAsJSON: {
+			cFieldsAsJSON:  tagsAsFKFieldsAsJSONInit,
+			!cFieldsAsJSON: tagsAsFKInit,
+		},
+	},
+	!cTagsAsFK: {
+		cTagsAsJSON: {
+			cFieldsAsJSON:  tagsAndFieldsAsJSONInit,
+			!cFieldsAsJSON: tagsAsJSONInit,
+		},
+		!cTagsAsJSON: {
+			cFieldsAsJSON:  fieldsAsJSONInit,
+			!cFieldsAsJSON: vanillaColumns,
+		},
+	},
+}
+
+// Iterate through an array of 'metrics' visiting only those indexed by 'indices'
+// and depending on 'tagsAsFK', 'tagsAsJSON', and 'fieldsAsJSON' generate the
+// desired columns (their name, type and which role they play) for both the
+// main metrics table in the DB, and if tagsAsFK == true for the tags table.
+func makeTargetColumns(
```

Comment:
Something about the complexity of this function makes me think that an interface might be helpful here, where each combo of flags is a different implementation of the interface. But maybe that is overkill, it’s just that the function is a bit tough to digest.

I don’t have anything concrete to offer on that front though.

Reply:
I extracted two of the for loops as functions and it greatly reduced the strain of reading the function. Also moved it to an interface, so it can be mocked now. But I didn't know how to simplify the map[bool]map[bool]map[bool]func since we would just be substituting 8 functions for 8 structs